File: rust/nfts/src/event_handlers.rs
use chaindexing::states::{ContractState, Filters, Updates};
use chaindexing::{
    EventContext, EventHandler, HandlerResult, SideEffectContext, SideEffectHandler,
};

use crate::states::Nft;

pub struct TransferHandler;

#[chaindexing::augmenting_std::async_trait]
impl EventHandler for TransferHandler {
    fn abi(&self) -> &'static str {
        "event Transfer(address indexed from, address indexed to, uint256 indexed tokenId)"
    }

    async fn handle_event<'a, 'b>(&self, context: EventContext<'a, 'b>) -> HandlerResult {
        let event_params = context.get_event_params();

        // Extract each parameter as exactly specified in the ABI:
        // "event Transfer(address indexed from, address indexed to, uint256 indexed tokenId)"
        let _from = event_params.get_address_string("from");
        let to = event_params.get_address_string("to");
        let token_id = event_params.get_u32("tokenId");

        if let Some(existing_nft) =
            Nft::read_one(&Filters::new("token_id", token_id), &context).await?
        {
            let updates = Updates::new("owner_address", &to);
            existing_nft.update(&updates, &context).await?;
        } else {
            let new_nft = Nft {
                token_id,
                owner_address: to,
            };

            new_nft.create(&context).await?;
        }

        Ok(())
    }
}

pub struct TransferSideEffectHandler;

#[chaindexing::augmenting_std::async_trait]
impl SideEffectHandler for TransferSideEffectHandler {
    type SharedState = ();

    fn abi(&self) -> &'static str {
        "event Transfer(address indexed from, address indexed to, uint256 indexed tokenId)"
    }

    async fn handle_event<'a>(
        &self,
        context: SideEffectContext<'a, Self::SharedState>,
    ) -> HandlerResult {
        let token_id = context.get_event_params().get_u32("tokenId");

        context
            .enqueue_outbox(
                "nft-transfer-notification",
                &format!("token {token_id} moved"),
            )
            .await?;

        Ok(())
    }
}

File: typescript/nfts/src/event-handlers.ts
import {
  PureHandler,
  SideEffectHandler,
  PureHandlerContext,
  SideEffectHandlerContext,
} from '../../../../chaindexing-ts/chaindexing-core/src';
import { Nft } from './states';

export class TransferHandler implements PureHandler {
  abi(): string {
    return 'event Transfer(address indexed from, address indexed to, uint256 indexed tokenId)';
  }

  async handleEvent(context: PureHandlerContext): Promise<void> {
    const eventParams = context.getEventParams();

    // Extract each parameter as exactly specified in the ABI:
    // "event Transfer(address indexed from, address indexed to, uint256 indexed tokenId)"
    const from = eventParams.getAddressString('from');
    const to = eventParams.getAddressString('to');
    const tokenId = eventParams.getU32('tokenId');

    // For now, we'll always create or update (the TypeScript implementation doesn't have readOne yet)
    if (from === '0x0000000000000000000000000000000000000000') {
      // Mint: create new NFT
      const newNft = new Nft(tokenId, to);
      await newNft.create(context);
      console.log(`Minted new NFT ${tokenId} to ${to}`);
    } else {
      // Transfer: create a new record (in a full implementation we'd update existing)
      const transferredNft = new Nft(tokenId, to);
      await transferredNft.create(context);
      console.log(`Transferred NFT ${tokenId} from ${from} to ${to}`);
    }
  }
}

export class TransferSideEffectHandler implements SideEffectHandler<any> {
  abi(): string {
    return 'event Transfer(address indexed from, address indexed to, uint256 indexed tokenId)';
  }

  async handleEvent(context: SideEffectHandlerContext<any>): Promise<void> {
    const eventParams = context.getEventParams();
    const tokenId = eventParams.getU32('tokenId');
    const to = eventParams.getAddressString('to');

    // Side effect: Send notification
    await this.sendTransferNotification(tokenId, to);
    
    console.log(`Sent transfer notification for NFT ${tokenId} to ${to}`);
  }

  private async sendTransferNotification(tokenId: number, toAddress: string): Promise<void> {
    // Mock notification service - in real implementation this would call
    // external services like email, push notifications, webhooks, etc.
    console.log(`📧 Notification: NFT ${tokenId} has been transferred to ${toAddress}`);
    
    // Simulate async operation
    await new Promise(resolve => setTimeout(resolve, 100));
  }
} 